### PR TITLE
Slashgroups refactor and separation of commands based on roles

### PR DESCRIPTION
**Pull request by Hikari-Haru**
Separated the commands into "dalle", "gpt" and "system" commands, each with their own roles you can set for access to them.
g -> chat and gpt-chat -> converse
Added pycord-multicog
Had some issues with multicog erroring out if the slash command didn't have a guildid parameter, took a while to figure out that was the issue

See the commit messages for further detail.

I'm thinking that since settings are being locked behind admin roles, it could be an idea to add options to /chat for passing along some custom parameters which are more useful to change on the fly, like temperature or the penalties.

Feedback on things you don't like is appreciated

### CODE REVIEW

**Comment by Kav-K** [comment]
Hey @Hikari-Haru these changes look great and I'll fully review and merge soon

> I'm thinking that since settings are being locked behind admin roles, it could be an idea to add options to /chat for passing along some custom parameters which are more useful to change on the fly, like temperature or the penalties.

I don't expect this to be a thing for long, I'd like a way for a server owner to set all roles as able to use the bot, for example, if they omit "ALLOWED_ROLES", it'll allow all roles to use the command, that's something that I'll implement soon after fixing #41

File: cogs/gpt_3_commands_and_converser.py
Comment (Kav-K) [inline cogs/gpt_3_commands_and_converser.py:198]: Maybe this is a bit confusing to call it "chat", since it isn't exactly a chat and it's more so just a query, maybe we can keep it as /g, or /query, or /ask-gpt?

File: models/check_model.py
Comment (Kav-K) [inline models/check_model.py:6]: I will probably need you to change the way that the roles are retrieved soon.. unfortunately the current EnvService is riddled with bugs as per #41 and I'm going to be implementing a new way shortly, I apologize for the inconvenience, and I'll ping you when I've rolled out the new changes for that

File: cogs/draw_image_generation.py
Comment (Kav-K) [inline cogs/draw_image_generation.py:21]: I want to confirm the difference here between commands.Cog and discord.Cog, it seems I've been using the old version (commands.Cog), right?

**Review by Kav-K — CHANGES_REQUESTED** [review]

File: models/check_model.py
Comment (Kav-K) [reply models/check_model.py:6]: Nvm on this, I think .env stuff was fixed without many major changes, but I'd still like to wait for a confirmed resolution on #41 before we merge this

File: cogs/draw_image_generation.py
Comment (Hikari-Haru) [reply cogs/draw_image_generation.py:21]: Afaik discord.Cogs and discord.ext.commands.Cogs works in the same way. But since we're not using any discord.ext functionality we don't really need to keep it imported and can thus just change commands.Cog to discord.Cog. Not really an important change either way it goes

File: cogs/gpt_3_commands_and_converser.py
Comment (Hikari-Haru) [reply cogs/gpt_3_commands_and_converser.py:198]: That's a good suggestion, I think ask is good, so the full command would be /gpt ask

Comment (Kav-K) [reply cogs/gpt_3_commands_and_converser.py:198]: Are all the commands forced to have the command group name at first? Like for example do we have to do `/dalle draw` now? Is there any way to make the command just `/ask`?

Comment (Hikari-Haru) [reply cogs/gpt_3_commands_and_converser.py:198]: You can tab the command so if you type '/ask' and hit tab it'll change itself to '/gpt ask' and on mobile you just write 'askæ and click it like normal. We could relegate just the system commands to being in a group then keep the rest separate, but add the relevant checks to the 'slash_command' checks decorator. This is mainly what I wanted feedback on. Personally I find it fine, but I also see why it could be cumbersome

Comment (Kav-K) [reply cogs/gpt_3_commands_and_converser.py:198]: I see, I also personally find it fine because I use tab a lot myself as sort of an automatic reflex lol. I also think the benefit of these being grouped outweighs the extra word in the command and minor hassle, so I think we should proceed, but just to confirm, is there no other way to group commands without a forced prefix for the group name?

File: cogs/image_prompt_optimizer.py
Comment (Kav-K) [inline cogs/image_prompt_optimizer.py:51]: Let’s change this to just “optimize” since it’s in the dalle group now

File: cogs/draw_image_generation.py
Comment (Kav-K) [inline cogs/draw_image_generation.py:9]: Also, what does multicog actually do?

**Comment by Kav-K** [comment]
Also, I do also like your idea of being able to pass parameters to /ask for admin users, let’s task that as a separate issue we can track :)

File: cogs/gpt_3_commands_and_converser.py
Comment (Hikari-Haru) [reply cogs/gpt_3_commands_and_converser.py:198]: Not API wise afaik, there's probably some wizardly ways I'm not aware of to do it. I'm not that knowledgeable on it

Comment (Kav-K) [reply cogs/gpt_3_commands_and_converser.py:198]: Got it, we can proceed with this merge then, I like the changes. I’ll test locally tomorrow and merge— away from my computer right now

File: cogs/draw_image_generation.py
Comment (Hikari-Haru) [reply cogs/draw_image_generation.py:9]: The decorator adds the slash commands to a list group named from the string. Then in apply iterates over them and sets the slash command guild_id to the slash group guild_id, removes it from pending commands and adds the slash group as the parent of the slash command, also applies the original cog, or the slash group cog as fallback.

**Comment by Hikari-Haru** [comment]
> Also, I do also like your idea of being able to pass parameters to /ask for admin users, let’s task that as a separate issue we can track :)

I didn't mean for admin users, more as a way for regular gpt users to be able to pass parameters without having access to the settings command, as a one-shot change.

**Comment by Kav-K** [comment]
> > Also, I do also like your idea of being able to pass parameters to /ask for admin users, let’s task that as a separate issue we can track :)
> 
> I didn't mean for admin users, more as a way for regular gpt users to be able to pass parameters without having access to the settings command, as a one-shot change. I have an idea for it that I'll try later that should be simple enough

Got it, sounds good to me, thanks again for all your work here

File: cogs/draw_image_generation.py
Comment (Kav-K) [inline cogs/draw_image_generation.py:9]: What is `pycord`? I thought the package name was `py-cord`, or referenced as `discord`, `pycord` doesn't seem to exist?

image

**Review by Kav-K — CHANGES_REQUESTED** [review]

File: cogs/draw_image_generation.py
Comment (Kav-K) [reply cogs/draw_image_generation.py:9]: I'm assuming this is the package? 
image

Comment (Kav-K) [reply cogs/draw_image_generation.py:9]: Is there any way to achieve what you've done here without using another package?

Comment (Hikari-Haru) [reply cogs/draw_image_generation.py:9]: Haha, it's the name of the extension, see this https://github.com/Dorukyum/pycord-multicog you probably have to install requirements again since I added it there

Comment (Hikari-Haru) [reply cogs/draw_image_generation.py:9]: > Is there any way to achieve what you've done here without using another package?

Could implement the commands from the package straight into a .py file

Comment (Kav-K) [reply cogs/draw_image_generation.py:9]: I see, hmm, it's all good, lets keep it as a separate package so we don't have to manually update it if py-cord changes and that multicog stuff gets updated, what do you think @Hikari-Haru?

**Comment by Hikari-Haru** [comment]
Doesn't let me comment on the review, but could just copy the commands straight into a .py file and import it like usual, need to credit though

**Comment by Kav-K** [comment]
Ah weird, I'll need to fix that visibility soon, but in the thread I said:

I see, hmm, it's all good, lets keep it as a separate package so we don't have to manually update it if py-cord changes and that multicog stuff gets updated by the maintainer, what do you think @Hikari-Haru?

**Comment by Hikari-Haru** [comment]
Yeah I think so as well

File: README.md
Comment (Kav-K) [inline README.md:220]: If you don't mind, could you update the README to reflect the new command structure and resolve the conflicts we have? :)

**Review by Kav-K — CHANGES_REQUESTED** [review]

File: README.md
Comment (Hikari-Haru) [reply README.md:220]: It'll be in a good while if I do it, >17 hours, I only have my laptop for some days now and work from 7-17:30 😅

Comment (Kav-K) [reply README.md:220]: Ah no worries at all, I'll do it! Thank you :) Happy work day

**Pull request merged**
